### PR TITLE
storage: prohibit splitting on Range.StartKey for non-SQL ranges

### LINKED ISSUE #30396: storage: node panics with MVCCFindSplitKey returned start key of range
Running a restore of several tables (around 50gb I think) on my laptop, more than half an hour in, the node panics with this error. I wiped and tried again, but got the same result.

```
F180919 01:43:17.503592 23580 storage/replica_command.go:274  [n1,split,s1,r1047/1:/{Meta2/Table/…-System/}] MVCCFindSplitKey returned start key of range
goroutine 23580 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0xc0000c0200, 0xc0000c02a0, 0x8005800, 0x1a)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:972 +0xd4
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x872bc80, 0xc000000004, 0x80058e6, 0x1a, 0x112, 0xc018e07ec0, 0x5d)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:841 +0x80e
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x6c02080, 0xc01310b3e0, 0x4, 0x2, 0x0, 0x0, 0xc0193619c0, 0x1, 0x1)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:85 +0x2d5
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x6c02080, 0xc01310b3e0, 0x1, 0x4, 0x0, 0x0, 0xc0193619c0, 0x1, 0x1)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:69 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Fatal(0x6c02080, 0xc01310b3e0, 0xc0193619c0, 0x1, 0x1)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:181 +0x6c
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).adminSplitWithDescriptor(0xc01cc3dc00, 0x6c02080, 0xc01310b3e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:274 +0x536
github.com/cockroachdb/cockroach/pkg/storage.(*splitQueue).processAttempt(0xc000366120, 0x6c02080, 0xc01310b3e0, 0xc01cc3dc00, 0xc00a9ba000, 0x34, 0x40, 0x466fc32, 0x3)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/storage/split_queue.go:170 +0x3e4
github.com/cockroachdb/cockroach/pkg/storage.(*splitQueue).process(0xc000366120, 0x6c02080, 0xc01310b3e0, 0xc01cc3dc00, 0xc00a9ba000, 0x34, 0x40, 0x1, 0xc00042c480)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/storage/split_queue.go:127 +0x77
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).processReplica(0xc00048a500, 0x6c020c0, 0xc01cc79b30, 0xc01cc3dc00, 0x0, 0x0)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:744 +0x368
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).processLoop.func1.2(0x6c020c0, 0xc01cc79b30)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:627 +0xb8
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0005650e0, 0x6c020c0, 0xc01cc79b30, 0xc00bcc8330, 0x29, 0x0, 0x0, 0xc00033e460)
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/Users/david/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:319 +0x134
```

### PR DESCRIPTION
**Pull request by nvb**
Fixes #30396.

Before this change, we allowed splits on a range's StartKey as long as
that key had multiple MVCC versions. This change logically extends #24896
to fix this issue as well by preventing MVCCFindSplitKey from returning
the range's start key. In doing so, it also prevents MVCCFindSplitKey
from returning the first key in its range, even if it's not equal to the
range's StartKey. This is free to enforce and seems like the right thing
to do.

Release note: None

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
I confirmed that the backup from #30396 succeeds with this change.

**Review by benesch — APPROVED** [review]
Rare to see that the fix for something this obscure this actually makes the code cleaner. LGTM! Nice find.

Tiny nit: might be worth updating the big ol' comment in MVCCFindSplitKey to describe this edge case too. https://github.com/cockroachdb/cockroach/blob/0f3ee2b991d1c337630224a9511b5715ce7fc7c4/pkg/storage/engine/mvcc.go#L2687

File: pkg/storage/engine/mvcc.go [hunk 1]
```diff
@@ -2728,21 +2728,25 @@ func MVCCFindSplitKey(
...
+	var minSplitKey roachpb.Key
 	if _, _, err := keys.DecodeTablePrefix(it.UnsafeKey().Key); err == nil {
 		// The first key in this range represents a row in a SQL table. Advance the
 		// minSplitKey past this row to avoid the problems described above.
 		firstRowKey, err := keys.EnsureSafeSplitKey(it.Key().Key)
 		if err != nil {
 			return nil, err
 		}
-		minSplitKey = roachpb.RKey(firstRowKey.PrefixEnd())
+		minSplitKey = firstRowKey.PrefixEnd()
```
Comment (benesch) [inline pkg/storage/engine/mvcc.go]: Wait, don't you want to drop the `.PrefixEnd` now that you compare against `minSplitKey` with `<` instead of `<=`?

**Review by benesch — CHANGES_REQUESTED** [review]

File: pkg/storage/engine/mvcc.go (hunk 1 above)
Comment (benesch) [reply pkg/storage/engine/mvcc.go]: Oh, no, never mind. I confused myself. I do think that you're right that this is broken for interleaved tables.

**Comment by nvanbenschoten** [comment]
Ok, this is ready for another review. I addressed the interleaved row issue, so I think it's all set now. @jordanlewis might also be interested in taking a look at this.

**Review by a-robinson — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 4 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-:-LMscrfW1ejWhO0y1HGC:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/engine/engine.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZAW0EvkSLwTwEYgI:-LMsZAW0EvkSLwTwEYgJ:bdfe6xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/engine.go#L103)):*
> ```Go
> 	// the split is roughly targetSize bytes. The returned key will never be
> 	// chosen from the key ranges listed in keys.NoSplitSpans and will always
> 	// sort equal to or after minSplitKey.
> ```

Should the two commits be squashed together? It seems odd to do something and then immediately undo it.
___
*[pkg/storage/engine/mvcc.go, line 2695 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZNja-pqqD08L2OfH:-LMsZNja-pqqD08L2OfI:bxadf39) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2695)):*
> ```Go
> 	// In addition, we must never return a split key that falls within a table
> 	// row. (Rows in tables with multiple column families are comprised of
> 	// multiple keys, one key per column family.) However, we do allows a split
> ```

s/allows/allow
___
*[pkg/storage/engine/mvcc.go, line 2696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZY6v0tmw6rc8tk28:-LMsZY6v0tmw6rc8tk29:bcf5tzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2696)):*
> ```Go
> 	// row. (Rows in tables with multiple column families are comprised of
> 	// multiple keys, one key per column family.) However, we do allows a split
> 	// key that falls between a row and its interleaved rows.
> ```

Huh, TIL. It'd be nice if we had a preference against doing so other than in exceptional cases with massive rows in the non-interleaved table or a lot of data in the interleaved table, but that'd obviously be future work.

**Review by jordanlewis — APPROVED** [review]
:lgtm: on the interleaved table bit. Nice tests, too!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-:-LMsge6S5PuYmZC9XET7:b-wwph3g)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

**Review by benesch — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445)* status: :shipit: complete! 2 of 0 LGTMs obtained

**Review by nvb — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-:-LMyIkNPAPrpbwtCXcrB:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/storage/engine/engine.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZAW0EvkSLwTwEYgI:-LMyIQhB0XWX9_jhRhdJ:b-gexvpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/engine.go#L103)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Should the two commits be squashed together? It seems odd to do something and then immediately undo it.
</blockquote></details>

Yeah, there's an argument to be made that they should be squashed because there's a bit of churn in this interface, but they fix different bugs, so I think I'll keep them separate.

EDIT: I went through and removed the churn.
___
*[pkg/storage/engine/mvcc.go, line 2695 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZNja-pqqD08L2OfH:-LMyIbLk7DnqzIcmN6xN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2695)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/allows/allow
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 2696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZY6v0tmw6rc8tk28:-LMyIc5kCx1qajjZ-bL4:b2bt2f4) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2696)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Huh, TIL. It'd be nice if we had a preference against doing so other than in exceptional cases with massive rows in the non-interleaved table or a lot of data in the interleaved table, but that'd obviously be future work.
</blockquote></details>

Yeah, that's an interesting point. I wonder how useful a preference is without a guarantee though.

**Comment by nvanbenschoten** [comment]
Fixed lint warning.

bors r+

**Pull request merged**